File: apps/api/src/api/routes/catalog.py
from fastapi import APIRouter

from src.core.runtime import get_runtime_container

router = APIRouter(tags=['catalog'])


@router.get('/workspaces')
def list_workspaces() -> dict:
    container = get_runtime_container()
    return {'items': container.metadata.list_workspaces()}


@router.get('/datasets')
def list_datasets() -> dict:
    container = get_runtime_container()
    return {'items': container.metadata.list_datasets()}


@router.get('/datasets/{dataset_id}')
def get_dataset(dataset_id: str) -> dict:
    container = get_runtime_container()
    dataset = container.metadata.get_dataset(dataset_id)
    versions = container.metadata.list_dataset_versions(dataset_id)
    return {'item': dataset, 'versions': versions}


@router.get('/datasets/{dataset_id}/versions')
def list_dataset_versions(dataset_id: str) -> dict:
    container = get_runtime_container()
    return {'items': container.metadata.list_dataset_versions(dataset_id)}

File: apps/bff/src/services/catalog.ts
import type { DatasetDetail, DatasetItem, DatasetVersion, WorkspaceItem } from '../types.js'
import { platformFetch } from './platform.js'

export async function listWorkspaces() {
  const response = (await platformFetch('/workspaces')) as { items?: WorkspaceItem[] }
  return response.items ?? []
}

export async function listDatasets() {
  const response = (await platformFetch('/datasets')) as { items?: DatasetItem[] }
  return response.items ?? []
}

export async function getDataset(datasetId: string) {
  return (await platformFetch(`/datasets/${datasetId}`)) as DatasetDetail
}

export async function listDatasetVersions(datasetId: string) {
  const response = (await platformFetch(`/datasets/${datasetId}/versions`)) as { items?: DatasetVersion[] }
  return response.items ?? []
}